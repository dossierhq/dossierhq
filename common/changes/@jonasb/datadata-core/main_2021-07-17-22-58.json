{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Add OkResult.map() and ErrorType.isErrorType()",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}