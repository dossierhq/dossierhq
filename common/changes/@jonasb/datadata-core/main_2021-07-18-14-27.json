{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Configure importHelpers for ts, and add tslib dependency",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}