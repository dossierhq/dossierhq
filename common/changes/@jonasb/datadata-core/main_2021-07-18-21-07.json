{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Export transpiled JavaScript",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}