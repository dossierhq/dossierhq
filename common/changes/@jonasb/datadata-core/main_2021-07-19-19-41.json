{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Fix export configuration",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}