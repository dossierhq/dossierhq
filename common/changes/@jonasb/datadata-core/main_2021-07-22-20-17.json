{
  "changes": [
    {
      "packageName": "@jonasb/datadata-core",
      "comment": "Allow string in schema field specification",
      "type": "none"
    }
  ],
  "packageName": "@jonasb/datadata-core",
  "email": "jonas.b@gmail.com"
}