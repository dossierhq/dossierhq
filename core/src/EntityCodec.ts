import type { ErrorType, Result, SessionContext } from '.';
import { ensureRequired } from './Assertions';
import type { EntitiesTable, EntityVersionsTable } from './DbTableTypes';
import * as EntityFieldTypeAdapters from './EntityFieldTypeAdapters';
import { notOk, ok } from './ErrorResult';

export type EntityValues = Pick<EntitiesTable, 'uuid' | 'type' | 'name'> &
  Pick<EntityVersionsTable, 'data'>;

interface Entityish {
  id: string;
  _type: string;
  _name: string;
  [fieldName: string]: unknown;
}

export function decodeEntity(context: SessionContext, values: EntityValues): Entityish {
  const schema = context.instance.getSchema();
  const entitySpec = schema.getEntityTypeSpecification(values.type);
  if (!entitySpec) {
    throw new Error(`No entity spec for type ${values.type}`);
  }
  const entity: Entityish = {
    id: values.uuid,
    _type: values.type,
    _name: values.name,
  };
  if (values.data) {
    for (const [fieldName, fieldValue] of Object.entries(values.data)) {
      const fieldSpec = schema.getEntityFieldSpecification(entitySpec, fieldName);
      if (!fieldSpec) {
        throw new Error(`No field spec for ${fieldName} in entity spec ${values.type}`);
      }
      const fieldAdapter = EntityFieldTypeAdapters.getAdapter(fieldSpec);
      const decodedData = fieldAdapter.decodeData(fieldValue);
      entity[fieldName] = decodedData;
    }
  }
  return entity;
}

export function encodeEntity(
  context: SessionContext,
  entity: { _type: string; _name: string; [fieldName: string]: unknown },
  defaultValuesEncoded: Record<string, unknown> | null
): Result<{ type: string; name: string; data: Record<string, unknown> }, ErrorType.BadRequest> {
  const assertion = ensureRequired({ 'entity._type': entity._type, 'entity._name': entity._name });
  if (assertion.isError()) {
    return assertion;
  }

  const { _type: type, _name: name } = entity;

  const schema = context.instance.getSchema();
  const entitySpec = schema.getEntityTypeSpecification(type);
  if (!entitySpec) {
    return notOk.BadRequest(`Entity type ${type} doesnâ€™t exist`);
  }

  const result: { type: string; name: string; data: Record<string, unknown> } = {
    type,
    name,
    data: {},
  };
  for (const fieldSpec of entitySpec.fields) {
    const fieldAdapter = EntityFieldTypeAdapters.getAdapter(fieldSpec);
    if (fieldSpec.name in entity) {
      const data = entity[fieldSpec.name];
      result.data[fieldSpec.name] = fieldAdapter.encodeData(data);
    } else if (defaultValuesEncoded && fieldSpec.name in defaultValuesEncoded) {
      result.data[fieldSpec.name] = defaultValuesEncoded[fieldSpec.name]; // is already encoded
    }
  }
  return ok(result);
}
