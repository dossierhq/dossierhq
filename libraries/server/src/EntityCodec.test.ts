import type { AdminSchemaSpecification } from '@jonasb/datadata-core';
import { AdminSchema, FieldType, RichTextBlockType } from '@jonasb/datadata-core';
import { describe, expect, test } from 'vitest';
import { forTest } from './EntityCodec';

const { collectDataFromEntity } = forTest;

const schemaSpec: AdminSchemaSpecification = {
  entityTypes: [
    {
      name: 'EntityCodecFoo',
      adminOnly: false,
      fields: [
        { name: 'string', type: FieldType.String },
        { name: 'strings', type: FieldType.String, list: true },
        { name: 'location', type: FieldType.Location },
        { name: 'locations', type: FieldType.Location, list: true },
        { name: 'bar', type: FieldType.EntityType, entityTypes: ['EntityCodecBar'] },
        { name: 'bars', type: FieldType.EntityType, list: true, entityTypes: ['EntityCodecBar'] },
        { name: 'reference', type: FieldType.EntityType },
        { name: 'valueOne', type: FieldType.ValueType },
        { name: 'richText', type: FieldType.RichText, entityTypes: ['EntityCodecBar'] },
        { name: 'richTexts', type: FieldType.RichText, list: true },
      ],
    },
    {
      name: 'EntityCodecBar',
      adminOnly: false,
      fields: [],
    },
  ],
  valueTypes: [
    {
      name: 'EntityCodecValueOne',
      adminOnly: false,
      fields: [
        { name: 'string', type: FieldType.String },
        { name: 'strings', type: FieldType.String, list: true },
        { name: 'location', type: FieldType.Location },
        { name: 'locations', type: FieldType.Location, list: true },
        { name: 'bar', type: FieldType.EntityType, entityTypes: ['EntityCodecBar'] },
        { name: 'richText', type: FieldType.RichText, entityTypes: ['EntityCodecBar'] },
        { name: 'child', type: FieldType.ValueType, valueTypes: ['EntityCodecValueOne'] },
      ],
    },
  ],
};

const schema = new AdminSchema(schemaSpec);

describe('collectDataFromEntity', () => {
  test('empty', () => {
    expect(collectDataFromEntity(schema, { info: { type: 'EntityCodecFoo' }, fields: {} }))
      .toMatchInlineSnapshot(`
        {
          "fullTextSearchText": [],
          "locations": [],
          "requestedReferences": [],
        }
      `);
  });

  test('name only', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {},
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [],
        "locations": [],
        "requestedReferences": [],
      }
    `);
  });

  test('strings', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {
          string: 'Hello string world',
          strings: ['one', 'two', 'three'],
        },
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [
          "Hello string world",
          "one",
          "two",
          "three",
        ],
        "locations": [],
        "requestedReferences": [],
      }
    `);
  });

  test('value item strings', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {
          valueOne: {
            type: 'EntityCodecValueOne',
            string: 'one',
            strings: ['two', 'three'],
            child: { type: 'EntityCodecValueOne', string: 'four' },
          },
        },
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [
          "one",
          "two",
          "three",
          "four",
        ],
        "locations": [],
        "requestedReferences": [],
      }
    `);
  });

  test('rich text strings', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {
          richText: {
            blocks: [
              { type: RichTextBlockType.paragraph, data: { text: 'one one' } },
              {
                type: RichTextBlockType.valueItem,
                data: { type: 'EntityCodecValueOne', string: 'one two' },
              },
              {
                type: 'header',
                data: { level: 3, text: 'Header text' },
              },
              {
                type: 'random',
                data: { foo: ['a', 'b'], bar: { z: 123 } },
              },
            ],
          },
          richTexts: [{ blocks: [{ type: RichTextBlockType.paragraph, data: { text: 'two' } }] }],
        },
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [
          "one one",
          "one two",
          "3",
          "Header text",
          "a",
          "b",
          "123",
          "two",
        ],
        "locations": [],
        "requestedReferences": [],
      }
    `);
  });

  test('entity locations', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {
          location: { lat: 1, lng: 2 },
          locations: [
            { lat: 3, lng: 4 },
            { lat: 5, lng: 6 },
          ],
        },
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [],
        "locations": [
          {
            "lat": 1,
            "lng": 2,
          },
          {
            "lat": 3,
            "lng": 4,
          },
          {
            "lat": 5,
            "lng": 6,
          },
        ],
        "requestedReferences": [],
      }
    `);
  });

  test('value item locations', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {
          valueOne: { type: 'EntityCodecValueOne', location: { lat: 1, lng: 2 } },
          richText: {
            blocks: [
              {
                type: RichTextBlockType.valueItem,
                data: { type: 'EntityCodecValueOne', location: { lat: 3, lng: 4 } },
              },
            ],
          },
        },
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [],
        "locations": [
          {
            "lat": 1,
            "lng": 2,
          },
          {
            "lat": 3,
            "lng": 4,
          },
        ],
        "requestedReferences": [],
      }
    `);
  });

  test('entity references', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {
          bar: { id: 'barId1' },
          bars: [{ id: 'barId2' }, { id: 'barId3' }],
          reference: { id: 'unspecifiedId1' },
        },
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [],
        "locations": [],
        "requestedReferences": [
          {
            "entityTypes": [
              "EntityCodecBar",
            ],
            "prefix": "entity.fields.bar",
            "uuids": [
              "barId1",
            ],
          },
          {
            "entityTypes": [
              "EntityCodecBar",
            ],
            "prefix": "entity.fields.bars[0]",
            "uuids": [
              "barId2",
            ],
          },
          {
            "entityTypes": [
              "EntityCodecBar",
            ],
            "prefix": "entity.fields.bars[1]",
            "uuids": [
              "barId3",
            ],
          },
          {
            "entityTypes": undefined,
            "prefix": "entity.fields.reference",
            "uuids": [
              "unspecifiedId1",
            ],
          },
        ],
      }
    `);
  });

  test('value item references', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {
          valueOne: { type: 'EntityCodecValueOne', bar: { id: 'bar1Id' } },
        },
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [],
        "locations": [],
        "requestedReferences": [
          {
            "entityTypes": [
              "EntityCodecBar",
            ],
            "prefix": "entity.fields.valueOne.bar",
            "uuids": [
              "bar1Id",
            ],
          },
        ],
      }
    `);
  });

  test('rich text reference', () => {
    expect(
      collectDataFromEntity(schema, {
        info: { type: 'EntityCodecFoo' },
        fields: {
          richText: {
            blocks: [
              { type: RichTextBlockType.entity, data: { id: 'barId1' } },
              {
                type: RichTextBlockType.valueItem,
                data: { type: 'EntityCodecValueOne', bar: { id: 'bar2Id' } },
              },
            ],
          },
        },
      })
    ).toMatchInlineSnapshot(`
      {
        "fullTextSearchText": [],
        "locations": [],
        "requestedReferences": [
          {
            "entityTypes": [
              "EntityCodecBar",
            ],
            "prefix": "entity.fields.richText[0]",
            "uuids": [
              "barId1",
            ],
          },
          {
            "entityTypes": [
              "EntityCodecBar",
            ],
            "prefix": "entity.fields.richText[1].bar",
            "uuids": [
              "bar2Id",
            ],
          },
        ],
      }
    `);
  });
});
