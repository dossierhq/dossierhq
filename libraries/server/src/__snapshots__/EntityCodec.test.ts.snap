// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`applySchemaMigrationsToFieldValues deleteField > nested value item 1`] = `
{
  "richText": {
    "root": {
      "children": [
        {
          "children": [
            {
              "detail": 0,
              "format": 0,
              "mode": "normal",
              "style": "",
              "text": "Heading 1",
              "type": "text",
              "version": 1,
            },
          ],
          "direction": "ltr",
          "format": "",
          "indent": 0,
          "tag": "h1",
          "type": "heading",
          "version": 1,
        },
        {
          "data": {
            "child": {
              "child": null,
              "type": "ValueItem",
            },
            "type": "ValueItem",
          },
          "format": "",
          "type": "valueItem",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
  },
}
`;

exports[`applySchemaMigrationsToFieldValues renameField > value item in rich text 1`] = `
{
  "richText": {
    "root": {
      "children": [
        {
          "children": [
            {
              "detail": 0,
              "format": 0,
              "mode": "normal",
              "style": "",
              "text": "Heading 1",
              "type": "text",
              "version": 1,
            },
          ],
          "direction": "ltr",
          "format": "",
          "indent": 0,
          "tag": "h1",
          "type": "heading",
          "version": 1,
        },
        {
          "data": {
            "child2": {
              "child2": null,
              "string2": "1.1",
              "type": "ValueItem",
            },
            "string2": "1",
            "type": "ValueItem",
          },
          "format": "",
          "type": "valueItem",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
  },
}
`;

exports[`applySchemaMigrationsToFieldValues renameType > value item in rich text 1`] = `
{
  "richText": {
    "root": {
      "children": [
        {
          "children": [
            {
              "detail": 0,
              "format": 0,
              "mode": "normal",
              "style": "",
              "text": "Heading 1",
              "type": "text",
              "version": 1,
            },
          ],
          "direction": "ltr",
          "format": "",
          "indent": 0,
          "tag": "h1",
          "type": "heading",
          "version": 1,
        },
        {
          "data": {
            "child": {
              "child": null,
              "string": "1.1",
              "type": "ValueItem2",
            },
            "string": "1",
            "type": "ValueItem2",
          },
          "format": "",
          "type": "valueItem",
          "version": 1,
        },
      ],
      "direction": "ltr",
      "format": "",
      "indent": 0,
      "type": "root",
      "version": 1,
    },
  },
}
`;
